### DISCUSSION TITLE
Flathub and JACK on Debian 11

### DISCUSSION THREAD

**Discussion by dyfer** [jacktrip/jacktrip#961] [Howtos] [answered]
I agree -- the Flathub build is very cool.  It loaded right up on Debian 11 and the GUI ran fine (first time in a long time for that).  

But here's an interesting thing.  I then went on to launch Ardour (which in turn launches Jack).  I then fired the Flathub Jacktrip as a Hub Server from the command line as I normally would.  Again, it ran fine. 

But it didn't see the running instance of jackd when I came in as a Hub client -- instead it came back with the "perhaps the JACK server isn't running?" error and quit.  But Jack was running.  Any ideas about that?   I've attached a screen grab.

![Flatpak install](https://user-images.githubusercontent.com/11680527/224568899-ff6a77d5-0ee9-47d8-9100-f70429fc1f9f.jpg)

_Originally posted by @OConnorStP in https://github.com/jacktrip/jacktrip/issues/952#issuecomment-1465282320_

**Comment by ntonnaett** [comment, marked as answer]
Flatpak's are sandboxed like apps in Android for example. Sandboxes increase the security a lot. If you open a compromised video or picture with your sandboxed viewer app, the effects outside of its sandbox are minimized. Your sandboxed social media app can't easily spy on your banking data. The downside to sandbox is that you need to break holes into it, because apps need to access the operating system, files, or devices. Pipewire is the only sound server that is able to work for sandboxed applications. So it's expected that the JackTrip Flatpak can't access your JACK server. Pipewire can be used as a JACK drop-in replacement. Every JACK client application should just work with Pipewire.

If you use Pipewire, you run into another problem with JackTrip because the default buffer size is 1024. You can set the latency by setting the environment variable `PIPEWIRE_LATENCY=128/48000`. But that's quite cumbersome. We might be able to add the feature to set the buffer size from the JackTrip gui like we do with the RtAudio backend. But the JACK API (programming interface) is very vague. If you set the buffer size via the JACK API when using the original JACK server, you change the buffer size for all application. With Pipewire this might only change the buffer size of JackTrip.

**Comment by ntonnaett** [comment]
What I've just realized. An interim way of doing this is to override `PIPEWIRE_LATENCY` for the JackTrip flatpak:
```
flatpak override --env=PIPEWIRE_LATENCY=128/48000 org.jacktrip.JackTrip
```

Not tested. But might be worse a shot. If it works we should document it.

**Comment by ntonnaett** [comment]
Pipewire will become default with Debian 12 at least for Gnome: https://www.phoronix.com/news/Debian-12-PipeWire

**Comment by ntonnaett** [comment]
Tested. The command must be slightly different: `flatpak override --user --env=PIPEWIRE_LATENCY=128/48000 org.jacktrip.JackTrip`

I think what I might do for the next release is to set the env variable to a default like 256/48000. So the default experience becomes much better but is still big enough to not introduce xruns. If people want lower latency they can override the variable.

**Comment by ntonnaett** [comment]
Setting a default value works too. This will make the default behavior of the Flatpak much better.

**Comment by ntonnaett** [comment]
I've created a PR for this for the current JackTrip release: https://github.com/flathub/org.jacktrip.JackTrip/pull/6
